ykrt/src/mt.rs

```diff
@@ -347,12 +355,76 @@ impl MTThread {
                     }
                 }
             }
-            PHASE_TRACING => {
+            PHASE_TRACING | PHASE_TRACING_LOCK => {
+                if lp & PHASE_TAG == PHASE_TRACING_LOCK {
+                    // PHASE_TRACING_LOCK requires special handling, because another thread might
+                    // have grabbed it on this Location even though this thread is tracing this
+                    // location.
+
+                    if self.inner.tracer.is_none() {
+                        // If this thread isn't tracing anything, there's little point in spinning
+                        // as the thread tracing this location is still plausibly alive.
+                        return None;
+                    }
+
+                    // We wait to grab the lock.
+                    // FIXME: this is a terrible spinlock.
+                    while lp & PHASE_TAG == PHASE_TRACING_LOCK {
```

Comment:
So here we know that this thread is tracing, but we couldn't have been the thread to put the location into `PHASE_TRACING_LOCK`?

Reply:
> So here we know that this thread is tracing

No, we don't know that: we just know that another thread is doing the same check as us.

Reply:
Gah! You're right.